caikit/runtime/client/remote_model_finder.py

```diff
@@ -0,0 +1,295 @@
+                    via the /info/models endpoint>
+                supported_models: Optional[Dict[str, str]]={} <mapping of model names to module_ids 
+                    that this remote supports. This is automatically populated by discovery_models>
+                    <model_path>: <module_id>
+
+"""
+# Standard
+from typing import Dict, Optional
+
+# Third Party
+import grpc
+import requests
+
+# First Party
+import aconfig
+import alog
+
+# Local
+from caikit.core.exceptions import error_handler
+from caikit.core.model_management.factories import model_finder_factory
+from caikit.core.model_management.model_finder_base import ModelFinderBase
+from caikit.interfaces.common.data_model.remote import ConnectionInfo
+from caikit.interfaces.runtime.data_model import ModelInfoRequest, ModelInfoResponse
+from caikit.runtime.client.remote_config import RemoteModuleConfig
+from caikit.runtime.names import (
+    MODEL_MESH_MODEL_ID_KEY,
+    MODELS_INFO_ENDPOINT,
+    ServiceType,
+    get_grpc_route_name,
+)
+
+log = alog.use_channel("RFIND")
+error = error_handler.get(log)
+
+
+### Finder Definitions
+
+
+class RemoteModelFinder(ModelFinderBase):
+    __doc__ = __doc__
+
+    name = "REMOTE"
+
+    def __init__(self, config: aconfig.Config, instance_name: str):
+        """Initialize with an optional path prefix"""
+
+        self._instance_name = instance_name
+
+        # Type/Value check connection parameters
+        self._connection = ConnectionInfo(**config.connection)
+        self._tls = self._connection.tls
+
+        # Type/Value check default parameters
+        self._model_key = config.get("model_key", MODEL_MESH_MODEL_ID_KEY)
+        error.type_check("<COR72281587E>", str, model_key=self._model_key)
+
+        self._protocol = config.get("protocol", "grpc")
+        error.value_check(
+            "<COR72281545E>",
+            self._protocol in ["grpc", "http"],
+            "Unknown protocol: %s",
+            self._protocol,
+        )
+
+        if self._protocol == "grpc" and self._tls.enabled:
+            error.value_check(
+                "<COR74451567E>",
+                not self._tls.insecure_verify,
+                "GRPC does not support insecure TLS connections."
+                "Please provide a valid CA certificate",
+            )
+
+        # Type/Value check model parameters
+        self._discover_models = config.get("discover_models", True)
+        self._supported_models = config.get("supported_models", {})
+        error.type_check(
+            "<COR74343245E>",
+            dict,
+            supported_models=self._supported_models,
+        )
+        error.type_check(
+            "<COR74342245E>",
+            bool,
+            discover_models=self._discover_models,
+        )
+
+        # Discover remote models
+        if self._discover_models:
+            self._supported_models.update(self.discover_models())
```

Comment:
Hmmm, that's a good point. After thinking about it some more how does this sound:
1. Discover models on startup
2. Discover models when model was not located in `self._supported_models`. To get around re-discovering models frequently we'd add a "last_discovered" variable which tracks the last time discovery was run and we only run it once every couple of minutes.

I didn't want to utilize a Cron since discovered models only matter during load and shouldn't need to be updated in the background.

Reply:
I like it! I agree that it's probably best to avoid a cron

Reply:
Ok, the new discovery logic looks great: Resilient without the need for a cron! A couple of tiny naming NITs, but I'm happy with where it is.